### ISSUE TITLE
swapon failed: Input/output error

### ISSUE THREAD

**Issue by macmpi** [maxnet/berryboot#408]
Hi,

I'm trying to use [Volumio2 distribution](https://volumio.org/forum/multiboot-volumio2-with-kodi-under-berryboot-t6818.html#p33742) under Berryboot.
It's mostly fine, except on lower-RAM devices when Volumio tries to dynamically enable a swap file with [this](https://github.com/volumio/Build/blob/master/volumio/bin/dynswap.sh#L20): this generates the error message in the title.
Tried manually, and same thing happens at last `swapon` command, verbose message says:
```
volumio@volumio:/data$ sudo swapon /data/swapfile --verbose
swapon /data/swapfile
swapon: /data/swapfile: found swap signature: version 1d, page-size 4, same byte order
swapon: /data/swapfile: pagesize=4096, swapsize=536870912, devsize=536870912
swapon: /data/swapfile: swapon failed: Input/output error
```
Could this be due to `aufs` filesystem ?
Any suggested workaround?

Thanks!

**Comment by maxnet** [comment]
You will need a swap partition instead of file if you really want that.
Do not recommend having swap at all though.

**Comment by macmpi** [comment]
Curious why it works on native, and not under berryboot though...
I'd like to minimize changes to image.
Maybe something [around this](https://sourceforge.net/p/aufs/mailman/message/23085077/) may do the trick...

**Comment by maxnet** [comment]
But why do you want to have swap in the first place?
SD cards do not make good swap devices.

**Comment by macmpi** [comment]
It's not me, it what's built-into the original distribution...
I doubt I have enough argument to push such kind of change in that proven distribution, built for many platforms (some not relying on SD for storage).
Some low-RAM devices (such as PiZero) are a bit short to run without swap apparently...

I guess some other Raspi distributions may use swap too, no?

**Comment by macmpi** [comment]
It seems RetroPie [does similar things](https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/helpers.sh#L304)

**Comment by maxnet** [comment]
Yeah, standard Raspbian has that as well, but it is the first thing one normally disables.
Your mouse pointer shocks every time it swaps if you have a cheap SD card.

**Comment by macmpi** [comment]
sure, but what if distribution does not work without it...
Any idea on how to have berryboot support this, if user can live with slowed (albeit working) app?
Thanks for any tip!

**Comment by maxnet** [comment]
Would need to activate a swap file before the aufs layer goes on, so in berryboot-init.
And even then it may not work if LUKS encryption was activated during Berryboot installation. In that case you can only use swap partitions, not files.

**Comment by macmpi** [comment]
Interesting, thanks, will look into that: any suggested location for that file at that time?
no LUKS for me ;p

**Comment by macmpi** [comment]
`berryboot-init` file must pertain to original squashfs image, so that it can be loaded from `/squashfs` directory when image is mounted, right?

Would be nice to have similar modifiable `berryboot-init` which may reside outside of squash image (could be in `/mnt/data` for instance), but with image name being passed as parameter, so that some switch case can be done according to specific images (like currently in a way).
This would avoid to unsquash/modify/resquash images for `berryboot-init` file inclusion at each image update...

Make sense?
A bit like similar [suggested change](https://github.com/maxnet/berryboot/issues/400#issuecomment-305009960) to enable custom image update scripts...

**Comment by macmpi** [comment]
Followed your suggestion (thanks !) and it seems to sort-of work.
However, while I created swap file as `/mnt/swpafile`, it gets remapped elsewhere after `aufs` magics and image mounting stuff (in `/media/berryboot/swapfile` or whatever), and original location does not exist anymore when image runs...
Still, `cat /proc/swaps` reports the original location and `free` shows swap is active/used !...
I wonder if this remapping-thing will not cause problems: it looks a bit weird.
Any suggestion on where it shall be set in first place, or eventually bind-mounted? Not a specialist of this...
Thanks.

PS: appreciate if you can consider the PR (or some better thought-out alternative?), which would definitely make things easier for `berryboot-init` feature.

**Comment by maxnet** [comment]
>Any suggestion on where it shall be set in first place, or eventually bind-mounted?

The only good place is in a partition.

But best is not to have it all. Ever.
I do not think swap is ever a good solution. Let alone on flash memory.
If your device has insufficient memory to do without it, get a different one.

**Comment by macmpi** [comment]
...or optimize RAM allocation (gpu vs arm) to set arm side to maximum, which would greatly help in that case.
But it seems there is an issue (https://github.com/maxnet/berryboot/issues/411) in setting `gpu_mem` through image file naming since move to 4.9 kernel...

**Comment by shunobies** [comment]
I had to create a new image for the most recent Raspberry Pi OS 64bit and had several issues with it that were solved by:
sudo mount /dev/mmcblk0p1 /boot

sudo crontab -e

@reboot mount /dev/mmcblk0p1 /boot

I wouldn't use the swap unless you have an SSD or NVME drive. If your using SD card this probably isn't a good idea. However the above commands allowed me to actual install some of the apt packages I was having issues with.

BerryBoot Workaround SwapFile
sudo nano /etc/dphys-swapfile

CONF_SWAPFILE=/boot/swapfile

sudo dphys-swapfile setup
sudo dphys-swapfile swapon
